### ISSUE TITLE
Format markdown documents containing Rust

### ISSUE THREAD

**Issue by Michael-F-Bryan** [rust-lang/rustfmt#2036]
I was wanting to do something similar to #1695, although in my case I've got a bunch of [markdown files] containing Rust snippets which I'd like to format.

I knocked up a quick Python script which *almost* does this ([gist](https://gist.github.com/Michael-F-Bryan/b3319588b017a7f906a0e4315ff5005e)).

However it looks like rustfmt will skip any file not ending in `*.rs`... Would it be possible to allow non-Rust files if they are explicitly specified using the `--file-lines` argument?

Output of running rustfmt from Python:

```
python rustfmt.py 

b"Warning: Extra file listed in file_lines option '/home/michael/Documents/playground/static-analyser-in-rust/src/codemap.md'\nWarning: Extra file listed in file_lines option '/home/michael/Documents/playground/static-analyser-in-rust/src/lib.md'\nWarning: Extra file listed in file_lines option '/home/michael/Documents/playground/static-analyser-in-rust/src/parse.md'\nWarning: Extra file listed in file_lines option '/home/michael/Documents/playground/static-analyser-in-rust/src/lowering.md'\nWarning: Extra file listed in file_lines option '/home/michael/Documents/playground/static-analyser-in-rust/src/analysis.md'\nWarning: Extra file listed in file_lines option '/home/michael/Documents/playground/static-analyser-in-rust/src/errors.md'\nWarning: Extra file listed in file_lines option '/home/michael/Documents/playground/static-analyser-in-rust/src/lex.md'\n\n"
b''
```

[markdown files]: https://github.com/Michael-F-Bryan/static-analyser-in-rust

**Comment by nrc** [comment]
Rustfmt formats Rust *programs* not Rust *files* - it runs according to the declared module structure. Therefore, running on markdown files might be a little complicated and I'm not sure how to do it. However, it definitely should be possible - it is basically the same idea as testing snippets in markdown.

**Comment by Michael-F-Bryan** [comment]
Ah that makes sense given how `rustfmt-nightly` uses bits of the compiler for parsing and analysis. So the process is a little more complicated than just "here's a string containing some Rust code, please format it".

Would there be any prospects for formatting markdown files or doc-comment examples in the future?

**Comment by nrc** [comment]
> Would there be any prospects for formatting markdown files or doc-comment examples in the future?

Yeah, it seems like something that it would be good to do and I'm pretty sure it is possible, I'm just not sure how to do it :-)

**Comment by hsribei** [comment]
Hi! I'm helping with the translation of TRPL to [Brazilian Portuguese](https://github.com/rust-br/rust-book-pt-br) (pt-BR) and I'm seeing demand for the same thing. There are many Rust snippets in markdown source that have to be translated, some symbols are translated in batch with search/replace, and it would be nice to run one command to rectify substitutions that have caused formatting mistakes. 

Of course this should be super useful to anyone producing Rust content, so the Content Team should be all over this =D Is this the right place to upvote/show support for the feature? Thank you!

**Comment by JelteF** [comment]
I would like to have this same feature. In my case it's the separate `.md` doc files that I have that describe the usage of my library. For instance this one: https://github.com/JelteF/derive_more/blob/master/doc/add.md

**Comment by JeanMertz** [comment]
I too would like to see a feature like this.

It would help with things like this:

https://github.com/rustic-games/things/commit/ae5280d71af606901354d0f62a9f5fe61a25ddfc

Which uses the (unstable) [`external_doc` feature](https://doc.rust-lang.org/beta/unstable-book/language-features/external-doc.html).

Although, I guess, this would be a bit more complicated, as there are references to modules that only make sense once that external code gets inlined into the main code itself.

Still though, having this would bring this "full circle", allowing you to externally document your library, while still getting proper formatting in those external files.

**Comment by lambda-fairy** [comment]
I've been discussing this with @hukkinj1 for Maud: lambda-fairy/maud#231

The conversation is focused on mdformat, which is written in Python. But I can imagine a hypothetical Markdown formatter using [pulldown-cmark](https://crates.io/crates/pulldown-cmark) that does the same thing.

**Comment by shepmaster** [comment]
My quick-and-dirty hack is to:

1. create a throwaway Cargo project
1. enable `format_code_in_doc_comments` in that project
1. copy the markdown file into the project's lib.rs as documentation for a function
1. format the library
1. strip out the documentation comments and function and copy it back to the original file  

This _seems_ to work OK. 

<details>
<summary>Script to format markdown</summary>

```bash
#!/usr/bin/env bash

set -euo pipefail

scratch_dir=$(mktemp -d)
project_dir="${scratch_dir}/scratch"

cargo_toml="${project_dir}/Cargo.toml"
rustfmt_toml="${project_dir}/rustfmt.toml"
lib_rs="${project_dir}/src/lib.rs"

cargo new --lib "${project_dir}"

cat <<-EOF > "${rustfmt_toml}"
format_code_in_doc_comments = true
EOF

for markdown_file in $(git ls-files | rg '.md$'); do
    awk '{ print "/// " $0 } END { print "fn dummy() {}"}' "${markdown_file}" > "${lib_rs}"
    cargo +nightly fmt --manifest-path="${cargo_toml}"
    sed -E -e '$ d' -e 's@/// ?@@' "${lib_rs}" > "${markdown_file}"
done
```

</details>

**Comment by mumbleskates** [comment]
Also works, doesn't write any files:

```bash
cat README.md | (
  while IFS= read -r LINE; do
    if [[ -z ${LINE} ]]; then
      echo "///"
    else
      echo "/// ${LINE}"
    fi
  done
  echo "fn dummy() {}"
) | rustfmt --check --config unstable_features=true,format_code_in_doc_comments=true,max_width=80
```
Notably `rustfmt` still returns a zero exit code even if it wants to make changes, so you should check that there is no output if this is somehow part of your CI.

**Comment by djukyo** [comment]
I would like to have this same feature.

Feature Request: Support Formatting Code Blocks in Markdown Files

Current Behavior:
The `cargo fmt` tool (powered by rustfmt) is designed to format Rust source files (.rs). However, it does not currently support formatting code blocks embedded within Markdown files.

Desired Behavior:
It would be beneficial if `cargo fmt` could directly format fenced code blocks in Markdown files, especially for Rust code. Additionally, supporting line-range formatting (e.g., `--line startline --endline endline`) would allow users to format specific portions of a file without affecting the entire document.

Use Case:
Many developers use Markdown for documentation and include Rust code snippets. Manually formatting these snippets can be time-consuming and error-prone. Extending `cargo fmt` to handle Markdown files would streamline this process.

Example Command:
```bash
cargo fmt --file README.md --line 10 --endline 20
```
